Setup ESLint using eslint-config-lisk-base - Closes #6

MaciejBaj opened the pull request:
Setup eslint basing on `eslint-config-lisk-base` configuration and apply suggested code changes on source and test files.
closes #6

MaciejBaj committed:
setup eslint and fix source files

MaciejBaj committed:
fix eslint errors in test files

MaciejBaj committed:
invoke static functions

willclarktech commented on .eslintrc:
```diff
@@ -0,0 +1,15 @@
+{
```
Best to call this file `.eslintrc.json` so IDEs/editors know how to parse it.

willclarktech commented on .eslintrc:
```diff
@@ -0,0 +1,15 @@
+    "beforeEach": true,
+    "after": true,
+    "afterEach": true,
+    "it": true
```
These globals are all added when you set `mocha` env to `true`.

MaciejBaj replied:
ok, thanks

willclarktech commented on .eslintrc:
```diff
@@ -0,0 +1,15 @@
+  },
+  "env": {
+    "node": true,
+    "mocha": true
```
It would be preferable to have your tests in their own directory and to set this `mocha` env in a nested `.eslintrc.json` file. That's probably a different issue though.

See e.g. https://github.com/LiskHQ/lisky/blob/development/test/.eslintrc.json

MaciejBaj replied:
I will set the `mocha` variable in each file instead, thanks

willclarktech commented on MasterWAMPServer.spec.js:
```diff
@@ -1,46 +1,47 @@
-'use strict';
+/* eslint no-unused-expressions: 0 */
+// https://github.com/jonathanglasmeyer/graphql-sequelize/commit/00814cac3aa9fa6d20aed38df838dcbc4b4ab9b4
```
This looks like probably the best solution here: https://github.com/prodatakey/dirty-chai

MaciejBaj replied:
ok, I will use dirty chai, thanks

willclarktech commented on MasterWAMPServer.spec.js:
```diff
@@ -1,46 +1,47 @@
 
-		it('should start listening on "workerStart"', function () {
+		it('should start listening on "workerStart"', () => {
+			/* eslint-disable no-new */
```
Prefer a rule with an exception for `expect`. See e.g. https://github.com/LiskHQ/eslint-config-lisk-base/blob/master/index.js#L8

MaciejBaj replied:
skipping the rule should be applied just for tests only. I will add the eslint variable setting at the beginning of tests files then

willclarktech commented on SlaveWAMPServer.js:
```diff
@@ -35,35 +33,43 @@ class SlaveWAMPServer extends WAMPServer {
 		});
 	}
 
+	static normalizeRequest(request = {}) {
```
👍

willclarktech commented on MasterWAMPServer.spec.js:
```diff
@@ -1,46 +1,47 @@
-const expect = chai.expect;
 
-describe('MasterWAMPServer', function () {
+const expect = chai.expect;
```
You can also do `const { expect } = require('chai')` directly.

MaciejBaj replied:
nice

MaciejBaj replied:
but now i have to use chai anyway to use `dirtyChai`: `chai.use(dirtyChai);` so I will leave it as it is

willclarktech replied:
If you're going to use `dirtyChai` in every test file it might be neater to have a single setup file which does this, and adds `expect` to `global`.

willclarktech commented on SlaveWAMPServer.spec.js:
```diff
@@ -68,162 +69,161 @@ describe('SlaveWAMPServer', function () {
 
-		let validRequest, validSlaveToMasterRequest;
+		let validRequest;
+		let	validSlaveToMasterRequest;
```
Spacing

willclarktech commented on WAMPClient.js:
```diff
@@ -1,10 +1,9 @@
-"use strict";
+
 
```
Empty lines.

MaciejBaj replied:
thx

willclarktech commented on WAMPClient.js:
```diff
@@ -13,6 +12,7 @@ class WAMPClient {
 	}
 
 	static generateSignature(procedureCalls) {
+		/* eslint no-bitwise: ["error", { "int32Hint": true }] */
```
Prefer adding this as a rule or even submitting a PR to https://github.com/LiskHQ/eslint-config-lisk-base
Maybe we just want to allow bitwise operators in general?

willclarktech replied:
Or use `Math.floor` or `parseInt` or whatever.

MaciejBaj replied:
tests are showing it is faster but not so drastically
https://jsperf.com/math-round-vs-bitwise-op/1

MaciejBaj replied:
I will use Math.ceil instead

willclarktech commented on WAMPClient.spec.js:
```diff
@@ -12,59 +13,53 @@ const WAMPResponseSchema = require('./schemas').WAMPResponseSchema;
-const clock = sinon.useFakeTimers(new Date(2020,1,1).getTime());
-
-describe('WAMPClient', function () {
+sinon.useFakeTimers(new Date(2020, 1, 1).getTime());
```
Wouldn't you rather reset the clock after the tests?

MaciejBaj replied:
good point, thanks

willclarktech commented on WAMPClient.spec.js:
```diff
@@ -12,59 +13,53 @@ const WAMPResponseSchema = require('./schemas').WAMPResponseSchema;
+	describe('wampSocket', () => {
+		describe('send', () => {
+			let wampClient;
+			let	wampSocket;
```
Spacing.

willclarktech commented on package.json:
```diff
@@ -7,7 +7,9 @@
-    "build:bundle": "BUNDLE=true webpack --optimize-minimize"
+    "build:bundle": "BUNDLE=true webpack --optimize-minimize",
+    "eslint": "eslint eslint **.js",
+    "eslint:fix": "eslint eslint --fix **.js"
```
`eslint eslint` in both of these commands

MaciejBaj replied:
uu

willclarktech commented on package.json:
```diff
@@ -22,6 +24,10 @@
+    "eslint": "^4.3.0",
+    "eslint-config-airbnb-base": "^11.3.1",
+    "eslint-config-lisk-base": "^0.1.0",
+    "eslint-plugin-import": "^2.7.0",
```
Isn't our practice to use exact versions?

MaciejBaj replied:
yes, it is :)

willclarktech commented on schemas.js:
```diff
@@ -1,6 +1,7 @@
-'use strict';
+
 
```
Empty lines.

MaciejBaj replied:
thx

willclarktech reviewed (changes requested):
In addition to the inline comments, I have 2 eslint errors and tests for `normalizeRequest` are failing.

MaciejBaj committed:
solve remaining eslint issues:
- use dirty chai for function style assertions
- set mocha variable for every file separately
- change WAMPServer.reply to static function
- use existing mocha globals
- fix libraries versions in package.json

MaciejBaj committed:
add eslint step to Jennkins configuration

willclarktech commented on Jenkinsfile:
```diff
@@ -17,6 +17,7 @@ def buildDependency() {
 def runTests() {
 	try {
 		sh '''#!/bin/bash
+		npm run eslint
```
Prefer a separate step so we can more easily tell whether lint or tests fail.

willclarktech commented on MasterWAMPServer.spec.js:
```diff
@@ -1,46 +1,46 @@
-'use strict';
-
+/* eslint-env node, mocha */
```
`node` is already specified in `.eslintrc.json` - does it need to be here also?

willclarktech commented on WAMPClient.spec.js:
```diff
@@ -258,3 +251,7 @@ describe('WAMPClient', function () {
+
+after(() => {
+	clock.restore();
+});
```
I much prefer to have these hooks all together at the start, but maybe we don't have a standard for this yet.

willclarktech commented on package.json:
```diff
@@ -7,27 +7,34 @@
-    "build:bundle": "BUNDLE=true webpack --optimize-minimize"
+    "build:bundle": "BUNDLE=true webpack --optimize-minimize",
+    "eslint": "eslint **.js",
+    "eslint:fix": "eslint --fix **.js"
```
These commands won't reach directories, e.g. `example`. (And there are lint errors in there.)

willclarktech reviewed (changes requested):
There are a couple of spacing issues still remaining from the previous review.

MaciejBaj committed:
use WAMPServer.reply as class function
reply function have to stay in WAMPServer instance as it being overwritten by MasterWAMPServer

MaciejBaj committed:
include example in linting, use test setup

MaciejBaj committed:
linting as a separate jenkins stage
